### PR TITLE
Test reporting improvements

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Comment by yuwata** [comment]
Not tested, but if I remember correctly, some `test-bus-*` are also silently skipped. If it is not included in this PR, I will submit follow-up commits about that.

**Comment by yuwata** [comment]
I notice that the following tests are still built:
```
test-dlopen
test-libsystemd-sym
test-libudev-sym
```

**Comment by yuwata** [comment]
@keszybz I've added some follow-ups here. If you do not like them, please drop them.

**Comment by keszybz** [comment]
@yuwata thanks. I changed your commit a bit to use `build_by_default` everywhere, and then added a follow-up commit to use that also in other places. This way all test binaries can be built on demand. PTAL.

**Comment by yuwata** [comment]
@keszybz Thanks. Sorry, my commits reference 8ecd4de but it is not included in this PR... Could you drop or update that?

> This way all test binaries can be built on demand.

Nice!

**Comment by poettering** [comment]
this needs a rebase, there are conflicts

File: src/shared/tests.c [hunk 1]
```diff
@@ -71,3 +72,10 @@ const char* get_testdata_dir(const char *suffix) {
         strncpy(testdir + strlen(testdir), suffix, sizeof(testdir) - strlen(testdir) - 1);
         return testdir;
 }
+
+bool slow_tests_enabled(void) {
+        int r;
+
+        r = getenv_bool("SYSTEMD_SLOW_TESTS");
+        return r >= 0 ? r : SYSTEMD_SLOW_TESTS_DEFAULT;
```
Comment (poettering) [inline src/shared/tests.c]: not that it matters, but i'd probably log here if r < 0, i.e. on parse errors...

**Comment by poettering** [comment]
lgtm

File: src/test/test-boot-timestamps.c [hunk 2]
```diff
@@ -91,5 +91,9 @@ int main(int argc, char* argv[]) {
         r = test_boot_timestamps();
         assert(r >= 0);
 
-        return (p > 0 || q > 0 || r >> 0) ? EXIT_SUCCESS : EXIT_TEST_SKIP;
+        bool any = p > 0 || q > 0 || r > 0;
```
Comment (filbranden) [inline src/test/test-boot-timestamps.c]: Note this seems to be fixing a typo here, `|| r >> 0)` in the preceding, while this one has `|| r > 0)`. This looks intentional to me. Probably good to add it to the commit description, just to make sure.

Also, I think code would be slightly better as:

```c
if (!(p > 0 || q > 0 || r > 0)) {   // or if (!any) also fine...
    log_notice("%s: no tests are possible.", program_invocation_short_name);
    return EXIT_TEST_SKIP;
}

return EXIT_SUCCESS;
```

Perhaps adding "skipping" to the log_notice message would also be good.

Comment (yuwata) [reply src/test/test-boot-timestamps.c]: > Perhaps adding "skipping" to the log_notice message would also be good.

:+1:

**Comment by yuwata** [comment]
Rebased. But the comments are not addressed.

**Comment by keszybz** [comment]
Updated, incorporating all comments.

File: src/libsystemd/sd-bus/test-bus-gvariant.c [hunk 3]
```diff
@@ -113,7 +114,7 @@ static void test_bus_gvariant_get_alignment(void) {
         assert_se(bus_gvariant_get_alignment("((t)(t))") == 8);
 }
 
-static void test_marshal(void) {
+static int test_marshal(void) {
         _cleanup_(sd_bus_message_unrefp) sd_bus_message *m = NULL, *n = NULL;
         _cleanup_(sd_bus_flush_close_unrefp) sd_bus *bus = NULL;
         _cleanup_free_ void *blob;
```
Comment (yuwata) [inline src/libsystemd/sd-bus/test-bus-gvariant.c]: Need to be initialized to NULL.

File: src/test/test-sleep.c [hunk 4]
```diff
@@ -26,10 +27,8 @@ static int test_fiemap(const char *path) {
         if (fd < 0)
                 return log_error_errno(errno, "failed to open %s: %m", path);
         r = read_fiemap(fd, &fiemap);
-        if (r == -EOPNOTSUPP) {
-                log_info("Skipping test, not supported");
-                exit(EXIT_TEST_SKIP);
-        }
+        if (r == -EOPNOTSUPP)
+                return log_tests_skipped("Not supported");
```
Comment (yuwata) [inline src/test/test-sleep.c]: Needs to check the return value in main().

Comment (keszybz) [reply src/test/test-sleep.c]: Fixed.

File: src/libsystemd/sd-bus/test-bus-gvariant.c (hunk 3 above)
Comment (keszybz) [reply src/libsystemd/sd-bus/test-bus-gvariant.c]: Good catch, thanks. Fixed now.

**Comment by keszybz** [comment]
Updated.

**Comment by yuwata** [comment]
lgtm.

**Comment by keszybz** [comment]
`bionic-s390x` — `boot-and-services    FAIL stderr: Failed to dump process list for 'systemd-nspawn@c1.service', ignoring: Unit systemd-nspawn@c1.service not loaded.`

**Comment by keszybz** [comment]
```
objdump: '/tmp/autopkgtest.I2I33x/build.SBH/debian/test/TEST-02-CRYPTSETUP/systemd': No such file
TEST RUN: cryptsetup systemd setup
+ /usr/bin/qemu-system-x86_64 -smp 1 -net none -m 512M -nographic -kernel /boot/vmlinuz-4.15.0-34-generic -drive format=raw,cache=unsafe,file=/var/tmp/systemd-test.rH1wwk/rootdisk.img -initrd /boot/initrd.img-4.15.0-34-generic -append 'ro root=/dev/sda1 raid=noautodetect loglevel=2 init=/lib/systemd/systemd console=ttyS0 selinux=0 printk.devkmsg=on   '
qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.vmx [bit 5]
Loading, please wait...
starting version 239
Begin: Loading essential drivers ... done.
Begin: Running /scripts/init-premount ... done.
Begin: Mounting root file system ... Begin: Running /scripts/local-top ... done.
Begin: Running /scripts/local-premount ... Scanning for Btrfs filesystems
done.
Warning: fsck not present, so skipping root file system
done.
Begin: Running /scripts/local-bottom ... done.
Begin: Running /scripts/init-bottom ... done.

Welcome to [1mUbuntu 18.04.1 LTS[0m!

[[0;32m  OK  [0m] Created slice system-systemd\x2dcryptsetup.slice.
[[0;32m  OK  [0m] Listening on initctl Compatibility Named Pipe.
[[0;32m  OK  [0m] Listening on Journal Audit Socket.
[[0;32m  OK  [0m] Listening on udev Kernel Socket.
[[0;32m  OK  [0m] Listening on Journal Socket (/dev/log).
[[0;1;33mUNSUPP[0m] Starting of Arbitrary Executable Fi…tem Automount Point not supported.
[[0;32m  OK  [0m] Reached target Swap.
[[0;32m  OK  [0m] Listening on Journal Socket.
         Mounting Kernel Debug File System...
         Starting File System Check on Root Device...
         Starting Journal Service...
         Mounting Huge Pages File System...
         Starting Load Kernel Modules...
[[0;32m  OK  [0m] Started Dispatch Password Requests to Console Directory Watch.
[[0;32m  OK  [0m] Created slice system-systemd\x2dfsck.slice.
[[0;32m  OK  [0m] Listening on udev Control Socket.
         Starting udev Coldplug all Devices...
[[0;32m  OK  [0m] Created slice User and Session Slice.
[[0;32m  OK  [0m] Reached target Slices.
[[0;32m  OK  [0m] Started Forward Password Requests to Wall Directory Watch.
[[0;32m  OK  [0m] Reached target Paths.
[[0;32m  OK  [0m] Created slice system-serial\x2dgetty.slice.
[[0;32m  OK  [0m] Listening on Process Core Dump Socket.
         Mounting POSIX Message Queue File System...
[[0;32m  OK  [0m] Mounted Kernel Debug File System.
[[0;32m  OK  [0m] Started File System Check on Root Device.
[[0;32m  OK  [0m] Mounted Huge Pages File System.
[[0;32m  OK  [0m] Started Load Kernel Modules.
[[0;32m  OK  [0m] Mounted POSIX Message Queue File System.
         Mounting FUSE Control File System...
         Starting Apply Kernel Variables...
         Mounting Kernel Configuration File System...
         Starting Remount Root and Kernel File Systems...
[[0;32m  OK  [0m] Mounted FUSE Control File System.
[[0;32m  OK  [0m] Started Apply Kernel Variables.
[[0;32m  OK  [0m] Mounted Kernel Configuration File System.
[[0;32m  OK  [0m] Started Journal Service.
[[0;32m  OK  [0m] Started Remount Root and Kernel File Systems.
         Starting Create System Users...
         Starting Rebuild Hardware Database...
[[0m[0;31m*     [0m] (1 of 6) A start job is running for…r-varcrypt.device (48s / no limit)
M
[K[[0;1;31m*[0m[0;31m*    [0m] (1 of 6) A start job is running for…r-varcrypt.device (49s / no limit)
M
[K[[0;31m*[0;1;31m*[0m[0;31m*   [0m] (1 of 6) A start job is running for…r-varcrypt.device (50s / no limit)
M
[K[ [0;31m*[0;1;31m*[0m[0;31m*  [0m] (2 of 6) A start job is running for…dplug all Devices (50s / no limit)
M
[K[  [0;31m*[0;1;31m*[0m[0;31m* [0m] (2 of 6) A start job is running for…dplug all Devices (51s / no limit)
M
[K[   [0;31m*[0;1;31m*[0m[0;31m*[0m] (2 of 6) A start job is running for…dplug all Devices (52s / no limit)
M
[K[[0;32m  OK  [0m] Started Create System Users.
[K         Starting Create Static Device Nodes in /dev...
[    [0;31m*[0;1;31m*[0m] (3 of 6) A start job is running for…dplug all Devices (58s / no limit)
M
[K[     [0;31m*[0m] (3 of 6) A start job is running for…dplug all Devices (59s / no limit)
M
[K[    [0;31m*[0;1;31m*[0m] (3 of 6) A start job is running for…plug all Devices (1min / no limit)
M
[K[[0;32m  OK  [0m] Started Create Static Device Nodes in /dev.
[K[[0;32m  OK  [0m] Reached target Local File Systems (Pre).
[[0;32m  OK  [0m] Started udev Coldplug all Devices.
[   [0;31m*[0;1;31m*[0m[0;31m*[0m] (4 of 4) A start job is running for…ware Database (1min 7s / 2min 17s)
M
[K[  [0;31m*[0;1;31m*[0m[0;31m* [0m] (4 of 4) A start job is running for…ware Database (1min 7s / 2min 17s)
M
[K[ [0;31m*[0;1;31m*[0m[0;31m*  [0m] (4 of 4) A start job is running for…ware Database (1min 8s / 2min 17s)
M
[K[[0;31m*[0;1;31m*[0m[0;31m*   [0m] (1 of 4) A start job is running for…rcrypt.device (1min 8s / no limit)
M
[K[[0;1;31m*[0m[0;31m*    [0m] (1 of 4) A start job is running for…rcrypt.device (1min 9s / no limit)
M
[K[[0m[0;31m*     [0m] (1 of 4) A start job is running for…rcrypt.device (1min 9s / no limit)
M
[K[[0;1;31m*[0m[0;31m*    [0m] (2 of 4) A start job is running for…4f745.device (1min 10s / 1min 30s)
M
[K[[0;31m*[0;1;31m*[0m[0;31m*   [0m] (2 of 4) A start job is running for…4f745.device (1min 11s / 1min 30s)
M
[K[ [0;31m*[0;1;31m*[0m[0;31m*  [0m] (2 of 4) A start job is running for…4f745.device (1min 11s / 1min 30s)
M
[K[  [0;31m*[0;1;31m*[0m[0;31m* [0m] (3 of 4) A start job is running for…ttyS0.device (1min 12s / 1min 30s)
M
[K[   [0;31m*[0;1;31m*[0m[0;31m*[0m] (3 of 4) A start job is running for…ttyS0.device (1min 13s / 1min 30s)
M
[K[    [0;31m*[0;1;31m*[0m] (3 of 4) A start job is running for…ttyS0.device (1min 13s / 1min 30s)
M
[K[     [0;31m*[0m] (4 of 4) A start job is running for…are Database (1min 14s / 2min 17s)
M
[K[    [0;31m*[0;1;31m*[0m] (4 of 4) A start job is running for…are Database (1min 14s / 2min 17s)
M
[K[   [0;31m*[0;1;31m*[0m[0;31m*[0m] (4 of 4) A start job is running for…are Database (1min 15s / 2min 17s)
M
[K[  [0;31m*[0;1;31m*[0m[0;31m* [0m] (1 of 4) A start job is running for…crypt.device (1min 15s / no limit)
M
[K[ [0;31m*[0;1;31m*[0m[0;31m*  [0m] (1 of 4) A start job is running for…crypt.device (1min 16s / no limit)
M
[K[[0;31m*[0;1;31m*[0m[0;31m*   [0m] (1 of 4) A start job is running for…crypt.device (1min 16s / no limit)
M
[K[[0;1;31m*[0m[0;31m*    [0m] (2 of 4) A start job is running for…4f745.device (1min 17s / 1min 30s)
M
[K[[0m[0;31m*     [0m] (2 of 4) A start job is running for…4f745.device (1min 18s / 1min 30s)
M
[K[[0;1;31m*[0m[0;31m*    [0m] (2 of 4) A start job is running for…4f745.device (1min 19s / 1min 30s)
M
[K[[0;31m*[0;1;31m*[0m[0;31m*   [0m] (3 of 4) A start job is running for…ttyS0.device (1min 19s / 1min 30s)
M
[K[ [0;31m*[0;1;31m*[0m[0;31m*  [0m] (3 of 4) A start job is running for…ttyS0.device (1min 20s / 1min 30s)
M
[K[  [0;31m*[0;1;31m*[0m[0;31m* [0m] (3 of 4) A start job is running for…ttyS0.device (1min 20s / 1min 30s)
M
[K[   [0;31m*[0;1;31m*[0m[0;31m*[0m] (4 of 4) A start job is running for…are Database (1min 20s / 2min 17s)
M
[K[    [0;31m*[0;1;31m*[0m] (4 of 4) A start job is running for…are Database (1min 21s / 2min 17s)
M
[K[     [0;31m*[0m] (4 of 4) A start job is running for…are Database (1min 21s / 2min 17s)
M
[K[    [0;31m*[0;1;31m*[0m] (1 of 4) A start job is running for…crypt.device (1min 22s / no limit)
M
[K[   [0;31m*[0;1;31m*[0m[0;31m*[0m] (1 of 4) A start job is running for…crypt.device (1min 23s / no limit)
M
[K[  [0;31m*[0;1;31m*[0m[0;31m* [0m] (1 of 4) A start job is running for…crypt.device (1min 23s / no limit)
M
[K[ [0;31m*[0;1;31m*[0m[0;31m*  [0m] (2 of 4) A start job is running for…4f745.device (1min 24s / 1min 30s)
M
[K[[0;31m*[0;1;31m*[0m[0;31m*   [0m] (2 of 4) A start job is running for…4f745.device (1min 25s / 1min 30s)
M
[K[[0;1;31m*[0m[0;31m*    [0m] (2 of 4) A start job is running for…4f745.device (1min 25s / 1min 30s)
M
[K[[0m[0;31m*     [0m] (3 of 4) A start job is running for…ttyS0.device (1min 26s / 1min 30s)
M
[K[[0;1;31m*[0m[0;31m*    [0m] (3 of 4) A start job is running for…ttyS0.device (1min 26s / 1min 30s)
M
[K[[0;31m*[0;1;31m*[0m[0;31m*   [0m] (3 of 4) A start job is running for…ttyS0.device (1min 27s / 1min 30s)
M
[K[ [0;31m*[0;1;31m*[0m[0;31m*  [0m] (4 of 4) A start job is running for…are Database (1min 27s / 2min 17s)
M
[K[  [0;31m*[0;1;31m*[0m[0;31m* [0m] (4 of 4) A start job is running for…are Database (1min 28s / 2min 17s)
M
[K[   [0;31m*[0;1;31m*[0m[0;31m*[0m] (4 of 4) A start job is running for…are Database (1min 28s / 2min 17s)
M
[K[    [0;31m*[0;1;31m*[0m] (1 of 4) A start job is running for…crypt.device (1min 29s / no limit)
M
[K[     [0;31m*[0m] (1 of 4) A start job is running for…crypt.device (1min 30s / no limit)
M
[K[    [0;31m*[0;1;31m*[0m] (1 of 4) A start job is running for…crypt.device (1min 30s / no limit)
M
[K[[0;1;31m TIME [0m] Timed out waiting for device dev-di…6d\x2d8b5e\x2d8b37c8b4f745.device.
[K[[0;1;33mDEPEND[0m] Dependency failed for Cryptography Setup for varcrypt.
[[0;1;33mDEPEND[0m] Dependency failed for dev-mapper-varcrypt.device.
[[0;1;33mDEPEND[0m] Dependency failed for File System Check on /dev/mapper/varcrypt.
[[0;1;33mDEPEND[0m] Dependency failed for /var.
[[0;1;33mDEPEND[0m] Dependency failed for Update UTMP about System Boot/Shutdown.
[[0;1;33mDEPEND[0m] Dependency failed for Update UTMP about System Runlevel Changes.
[[0;1;33mDEPEND[0m] Dependency failed for Flush Journal to Persistent Storage.
[[0;1;33mDEPEND[0m] Dependency failed for Local File Systems.
[[0;1;33mDEPEND[0m] Dependency failed for Load/Save Random Seed.
[[0;1;33mDEPEND[0m] Dependency failed for Local Encrypted Volumes.
[[0;32m  OK  [0m] Reached target Timers.
[[0;32m  OK  [0m] Reached target Login Prompts.
[[0;32m  OK  [0m] Reached target Sockets.
[[0;32m  OK  [0m] Started Emergency Shell.
[[0;32m  OK  [0m] Reached target Emergency Mode.
         Starting Commit a transient machine-id on disk...
         Starting Create Volatile Files and Directories...
[[0;32m  OK  [0m] Started Commit a transient machine-id on disk.
You are in emergency mode. After logging in, type "journalctl -xb" to view
system logs, "systemctl reboot" to reboot, "systemctl default" or "exit"
to boot into default mode.
Press Enter for maintenance
(or press Control-D to continue): autopkgtest [12:29:52]: ERROR: timed out on command "su -s /bin/bash root -c set -e; export USER=`id -nu`; . /etc/profile >/dev/null 2>&1 || true;  . ~/.profile >/dev/null 2>&1 || true; buildtree="/tmp/autopkgtest.I2I33x/build.SBH/debian"; mkdir -p -m 1777 -- "/tmp/autopkgtest.I2I33x/upstream-artifacts"; export AUTOPKGTEST_ARTIFACTS="/tmp/autopkgtest.I2I33x/upstream-artifacts"; export ADT_ARTIFACTS="$AUTOPKGTEST_ARTIFACTS"; mkdir -p -m 755 "/tmp/autopkgtest.I2I33x/autopkgtest_tmp"; export AUTOPKGTEST_TMP="/tmp/autopkgtest.I2I33x/autopkgtest_tmp"; export ADTTMP="$AUTOPKGTEST_TMP"; export DEBIAN_FRONTEND=noninteractive; export LANG=C.UTF-8; export DEB_BUILD_OPTIONS=parallel=1; unset LANGUAGE LC_CTYPE LC_NUMERIC LC_TIME LC_COLLATE   LC_MONETARY LC_MESSAGES LC_PAPER LC_NAME LC_ADDRESS   LC_TELEPHONE LC_MEASUREMENT LC_IDENTIFICATION LC_ALL;rm -f /tmp/autopkgtest_script_pid; set -C; echo $$ > /tmp/autopkgtest_script_pid; set +C; trap "rm -f /tmp/autopkgtest_script_pid" EXIT INT QUIT PIPE; cd "$buildtree"; export AUTOPKGTEST_NORMAL_USER=ubuntu; export ADT_NORMAL_USER=ubuntu; export 'CFLAGS=-O0'; export 'DEB_BUILD_PROFILES=noudeb'; export 'TEST_UPSTREAM=1'; export 'UPSTREAM_PULL_REQUEST=10070'; export 'GITHUB_STATUSES_URL=https://api.github.com/repos/systemd/systemd/statuses/730d989acc5cd913ffbf747570713c10f46516c4'; chmod +x /tmp/autopkgtest.I2I33x/build.SBH/debian/debian/tests/upstream; touch /tmp/autopkgtest.I2I33x/upstream-stdout /tmp/autopkgtest.I2I33x/upstream-stderr; /tmp/autopkgtest.I2I33x/build.SBH/debian/debian/tests/upstream 2> >(tee -a /tmp/autopkgtest.I2I33x/upstream-stderr >&2) > >(tee -a /tmp/autopkgtest.I2I33x/upstream-stdout);" (kind: test)
autopkgtest [12:29:52]: test upstream: -----------------------]
```
bionic-amd64 fails like this. I'll retrigger to see if this was an unrelated timeout.

**Comment by yuwata** [comment]
The TEST-02-CRYPTSETUP failure in bionic-amd64 was also found in #10062 (now it passed). So, I think it is not related to this. Note that #10047 includes this PR but the cryptsetup failure not found. Maybe, it randomly fails?

File: meson.build [hunk 5]
```diff
@@ -270,12 +270,14 @@ pkgconfig = import('pkgconfig')
...
-        cxx = find_program('c++', required : false)
-        if cxx.found()
-                #  Used only for tests
-                add_languages('cpp')
-        endif
+want_tests = get_option('tests')
+slow_tests = want_tests != 'false' and get_option('slow-tests')
+install_tests = get_option('install-tests')
+
+cxx = find_program('c++', required : false)
```
Comment (jameshilliard) [inline meson.build:277]: This doesn't work correctly for cross compilation:
```
Program c++ found: YES (/usr/bin/c++)

meson.build:293:8: ERROR:  Unknown compiler(s): [['/home/buildroot/buildroot/output/host/bin/i586-buildroot-linux-uclibc-g++']]
The follow exceptions were encountered:
Running "/home/buildroot/buildroot/output/host/bin/i586-buildroot-linux-uclibc-g++ --version" gave "[Errno 2] No such file or directory: '/home/buildroot/buildroot/output/host/bin/i586-buildroot-linux-uclibc-g++': '/home/buildroot/buildroot/output/host/bin/i586-buildroot-linux-uclibc-g++'"
```

Comment (yuwata) [reply meson.build:277]: Could you file this as a new issue?

Comment (jameshilliard) [reply meson.build:277]: Reported in #11278 now.

**Pull request merged**